// ionic-native-http-connection-backend/dist/native-http-module.js
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { HTTP } from '@ionic-native/http/ngx';
import { NativeHttpBackend } from './native-http-backend';
import * as i0 from "@angular/core";
var NativeHttpModule = /** @class */ (function () {
    function NativeHttpModule() {
    }
    NativeHttpModule.ɵmod = i0.ɵɵdefineNgModule({ type: NativeHttpModule });
    NativeHttpModule.ɵinj = i0.ɵɵdefineInjector({ factory: function NativeHttpModule_Factory(t) { return new (t || NativeHttpModule)(); }, providers: [HTTP, NativeHttpBackend], imports: [[HttpClientModule]] });
    return NativeHttpModule;
}());
export { NativeHttpModule };
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(NativeHttpModule, { imports: [HttpClientModule] }); })();
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(NativeHttpModule, [{
        type: NgModule,
        args: [{
                imports: [HttpClientModule],
                providers: [HTTP, NativeHttpBackend],
            }]
    }], null, null); })();
